### ISSUE TITLE
ecrecover return wrong values

### ISSUE THREAD

**Issue by legioner0** [ethereum/solidity#5109]
I found strange behavior of function `ecrecover` in solidity 0.4.24 and 0.4.25 - some input data leads to wrong return value.
Here is project with proof: https://github.com/legioner0/ecrecover-failed
In this project test working in `ganache`, but I started my investigation, when transactions to contract with similar logic on mainnet failed.

## Environment

- Compiler version: 0.4.24 and 0.4.25
- Framework/IDE (e.g. Truffle or Remix): truffle 4.1.11 and 4.1.14
- EVM execution environment / backend / blockchain client: ganache
- Operating system: Ubuntu 18.04.1 LTS amd64

**Comment by chriseth** [comment]
Thanks for the report! Unfortunately, the truffle test case is rather hard to follow. Can you provide the source code, the inputs, the actual outputs and the expected outputs? Are you able to reproduce the same execution in Remix?

**Comment by legioner0** [comment]
Ok, I updated my project. Now it has cli with deployment function. It may be useful for future investigation.

So, here https://ropsten.etherscan.io/address/0x4e146f2a85f51d73aa56cf84b1879e1c2fb056d5 is my test contract with verified source code.

Pkey used for sign is `0x57983e408e9a2cc43e5a9bf3285960bf0e79b1d06a667034d50e0a0ea01ef5dc` and `0xa411a48f77723c04ad91ec608edffc30151e50ea` it's corresponding address 

For user `0x9d4a66be4eefd7b66dde591755ceb4507afde0ef` signature is `{v: 0x1b, r: 0x5a7e6dd668a8810ad13f0faf031f78fd06cda414977b4c2efe866fbbd4959c59, s: 0x684ff3c2d00b5a736d4896ba9ecd6e864680ec6473ff252109ee94fd6a0f28f7}` and ecrecover in contract return correct value - `0xa411a48f77723c04ad91ec608edffc30151e50ea` (owner of pkey)

For user `0xaf335969601148d5927b08d440fefab6975da386` signature is `{v: 0x1b, r: 0xb00986d8bb52ee7acb06cabfa6c2c099d8904c7c8d56707a267ddbafd7aed070, s: 0x4068f5b5e6c4b442e83fcb7b6290520ebb5e077cd10d3bd86cf431ca4b640162}` and ecrecover in contract return incorrect value - `0x0000000000000000000000000000000000000000`. At same time ecrecover from `ethereumjs-util` return correct value - `0xa411a48f77723c04ad91ec608edffc30151e50ea`.

Transactions `0xdd69e9950f52dddcbc6751fdbb6949787cc1b84ac4020ab0617ec8ad950e554a` and `0x69e5a94fbedfc3529ed4978d18a418306f1356dd71fa2a535a2cc38f0fa98a7a` are show behavior, as described above.

**Comment by legioner0** [comment]
And third example:
For user `0x00131d69610d7ba5915f63e81db22b7c6a1e2d6e` signature is `{v: 0x1c, r: 0xb59a76f825b501cd2d116f8b672308a9a5f875e177f5ecf83bec004a06e235e0, s: 0x2fb9a7d44e7065cde9241c71816121277f5c9668abd0bb46a9edc7a764274180}` and ecrecover in contract return incorrect value - `0x39676acd4cd1482d43c3e69acfb1c7100c151ff6`. At same time ecrecover from `ethereumjs-util` return correct value - `0xa411a48f77723c04ad91ec608edffc30151e50ea.`

This behavior you may see in transaction `0xc7ecff0596acc1d7587e6e47b93d839baf9d6c2cf784e1175c323892057921f9`

**Comment by BingXinJiang** [comment]
I have this problem too

**Comment by chriseth** [comment]
Please dig deeper into this and check what the actual input data for the ecrecover precompile is. Perhaps this can help you: https://ethereum.stackexchange.com/questions/15364/ecrecover-from-geth-and-web3-eth-sign

This is very likely not an issue with the compiler.

**Comment by legioner0** [comment]
@chriseth, I know about this feature with prefixed hash. That's why I used `EthUtil.ecsign(EthUtil.keccak(Buffer.from(user.replace(/^0x/, ''), 'hex')), Buffer.from(pkey, 'hex'))` instead of `EthUtil.ecsign(EthUtil.hashPersonalMessage(EthUtil.keccak(Buffer.from(user.replace(/^0x/, ''), 'hex'))), Buffer.from(pkey, 'hex'))`. And also I tested my code - 11 of 100 input data variants failed the test, others 89 - complete successfully.

But as You insist, I updated my test project and deployed new contract to ropsten (also with verified source code) - https://ropsten.etherscan.io/address/0xc93a7d798552f238c8e5af41815894eb730f5cc9 Now js part and contract use prefixed hash.

Pkey used for sign is `0x57983e408e9a2cc43e5a9bf3285960bf0e79b1d06a667034d50e0a0ea01ef5dc` and `0xa411a48f77723c04ad91ec608edffc30151e50ea` it's corresponding address.

Here three examples for new version of contract: 
1. In transaction `0x8983bb0210b71720bea50115ae3f94b9fa4f4a2a05ad5e920d960621c2580961` I tried to sign prefixed message with address `0x7d67be226ec01d75d54e0ab0fe56a6316f710d85`. Signature is `{v: 0x1c, r: 0xd0bacdbcc8c2bb243683cf520958f60b7efbf3239c8df80db7e4d788f7e04886, s: 0x5a809be4e41e5b92ca284d4e0803544f6bc2d0ef9f57e0d1e2d5c0f4165614df}` and `ecrecover` in contract returns correct value - `0xa411a48f77723c04ad91ec608edffc30151e50ea`, same as returns value of `ercrecover` from `ethereumjs-util`.
2. In transaction `0xc0359df1b0720ef48a05c429353ed5adfa45ddebc12eeb790d5b29597d30694b` I tried to sign prefixed message with address `0x080f9e8de14b7efcc4918fd1eba4e1347ff9d319`. Signature is `{v: 0x1b, r: 0xa9262e11527b2df66c5b8f86894bcc0394a3892f2d93a1b6dd79cfdcc93bad35, s: 0x9ae9c34a55fe52fca41c23d428f68c3241902b2db31a586251d481f98d3bf130}` and `ecrecover` in contract returns incorrect value - `0xe46f54929281e8a5c68cbd35892db066ff57dbc8`. At the same time `ecrecover` from `ethereumjs-util` returns correct value - `0xa411a48f77723c04ad91ec608edffc30151e50ea`.
3. In transaction `0x51147f2b3659bd2f6d54bb0ac95e23452f5466fa67b5d55a240829500280d4d0` I tried to sign prefixed message with address `0x2f44bd8fb5bd72f5e35914d04f28d81acf727e6e`. Signature is `{v: 0x1c, r: 0x51147f2b3659bd2f6d54bb0ac95e23452f5466fa67b5d55a240829500280d4d0, s: 0x2f3af34df3d977850da5bc8be900eed0cb9873731c9c7783aa4be23cce0f36a6}` and `ecrecover` in contract returns incorrect value - `0x0000000000000000000000000000000000000000`. At the same time `ecrecover` from `ethereumjs-util` returns correct value - `0xa411a48f77723c04ad91ec608edffc30151e50ea`.

**Comment by chriseth** [comment]
If it fails for some inputs and succeeds for some others, the error is very very likely not in Solidity but rather in the VM implementation. Please compare the actual inputs to the ecrecover precompile with the actual inputs you supply to the other ecrecover implementation you use. Note that there are also different definitions of what a valid input for ecrecover is supposed to be.

**Comment by legioner0** [comment]
Oh, I found error, and it's much more simple.
This is incorrect converting result of `EthUtil.ecsign` to transaction data - I've lost leading zeros in bytes32 values and Web3 automatically repair length by adding trailing zeros. So `r` and `s` under some conditions changes
